dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,517 @@
+                "`input_names` should containi the same exact number of strings, one for "
+                "each pipeline input to be mapped by the proxy callable object"
+            )
+        self.num_inputs = len(self.dali_input_names)
+
+        # Multi-process queue used to transfer data from the pytorch workers to the main process
+        self.dali_input_q = mp.Queue()
+        # Multi-process queue used by the main process to consume outputs from the DALI pipeline
+        self.dali_output_q = queue.Queue()
+        # Thread
+        self.thread = None
+        # Cache
+        self.cache_outputs = dict()
+        # Whether we want the order of DALI execution to be reproducible
+        self.deterministic = deterministic
+
+    @property
+    def proxy(self):
+        return _DALIProxy(self.dali_input_names, self.dali_input_q, self.deterministic)
+
+    def _schedule_batch(self, pipe_run_ref):
+        torch.cuda.nvtx.range_push(f"dali_proxy.dali_input_q.put {pipe_run_ref.batch_id}")
+        self.dali_input_q.put((pipe_run_ref.batch_id, pipe_run_ref.inputs))
+        torch.cuda.nvtx.range_pop()
+
+    def _get_outputs(self, pipe_run_ref: DALIPipelineRunRef):
+        """
+        Gets the pipeline outputs for a specific batch id. It will keep reading data until the
+        right batch is found, caching the results that were not consumed until a future call
+        """
+        req_batch_id = pipe_run_ref.batch_id
+
+        # In case we haven't scheduled the iteration yet (i.e. deterministic config), do it now
+        if not pipe_run_ref.is_scheduled:
+            self._schedule_batch(pipe_run_ref)
+            pipe_run_ref.is_scheduled = True
+
+        # Wait for the requested output to be ready
+        req_outputs = None
+        # If the data was already read, just return it (and clear the cache entry)
+        if req_batch_id in self.cache_outputs:
+            req_outputs = self.cache_outputs[req_batch_id]
+            del self.cache_outputs[req_batch_id]
+
+        else:
+            curr_batch_id = None
+            # If not the data we are looking for, store it and keep processing until we find it
+            while req_batch_id != curr_batch_id:
+                torch.cuda.nvtx.range_push("dali_output_q.get")
+                curr_batch_id, curr_processed_outputs, err = self.dali_output_q.get()
+                torch.cuda.nvtx.range_pop()
+
+                if err is not None:
+                    raise err
+
+                if curr_batch_id == req_batch_id:
+                    req_outputs = curr_processed_outputs
+                else:
+                    self.cache_outputs[curr_batch_id] = curr_processed_outputs
+        # Unpack single element tuples
+        if isinstance(req_outputs, tuple) and len(req_outputs) == 1:
+            req_outputs = req_outputs[0]
+        return req_outputs
+
+    def produce_data(self, obj):
```

Comment:
Again, I have a question, so that I understand better. We can only finde one DALIPipelineRunRef here, right? If we found more in one "batch" it would be wrong?

Reply:
Following up on that, I think we may need to discuss how do we handle more than one output from the DALI pipeline. In theory, user should probably be able to get a tuple from the `__call__` on the proxy and than do stuff like return a dictionary:
```
processed_images, processed_labels = proxy(images, labels)
return {
    'images': processed_images,
    'labels': processed_labels,
    'text': some_text_thats_not_processed_by_dali
}
```
Than this whole `produce_data` would make sense.